### ISSUE TITLE
Add ability to pass build arguments via command line

### ISSUE THREAD

**Issue by dnestoro** [graalvm/native-build-tools#631]
**Is your feature request related to a problem? Please describe.**
Right now there is no way to pass build args to builds that are invoked via Maven, such as for Spring apps. Currently this can only be done via Maven profiles, writing which is a lot of overhead for passing arguments

**Describe the solution you'd like**
We need to provide a way to do something like `mvn -Pnative native:compile -DbuildArg=<...>`, so that arguments can be passed via command line. This should also support passing several arguments, andalso the argument `-o` for specifying the image name.

**Comment by melix** [comment]
In general this isn't such a good idea. Arguments passed by command line are by nature undocumented, opaque. Imagine build reproducibility: everything should be found in the build files, or it means that only god knows what was used to produce an artifact.

**Comment by olpaw** [comment]
> everything should be found in the build files

I agree. That's what we have the bundle feature for (https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/Bundles.md). If you want a self-contained file that allows you to replay a NI build, this should be your first choice.

BTW, the above is already possible even without adding support for `-DbuildArg=<...>` by using the `NATIVE_IMAGE_OPTIONS` environment variable. See https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md#picked-up-native_image_options

**Comment by alina-yur** [comment]
Hello. My use case is that if I want to customize the build process of the same application, such as for benchmarking, when I need an instrumented version, a plain Native Image version, and a Native Image version with all optimizations, this is what I have to do at the moment:

```xml
	<profiles>
		<profile>
			<id>instrumented</id>
			<build>
				<plugins>
					<plugin>
						<groupId>org.graalvm.buildtools</groupId>
						<artifactId>native-maven-plugin</artifactId>
						<configuration>
							<imageName>demo-instrumented</imageName>
							<buildArgs>
								<buildArg>--pgo-instrument</buildArg>
							</buildArgs>
						</configuration>
					</plugin>
				</plugins>
			</build>
		</profile>
		<profile>
			<id>optimized</id>
			<build>
				<plugins>
					<plugin>
						<groupId>org.graalvm.buildtools</groupId>
						<artifactId>native-maven-plugin</artifactId>
						<configuration>
							<imageName>demo-optimized</imageName>
							<buildArgs>
								<buildArg>--gc=G1</buildArg>
								<buildArg>--pgo=${project.basedir}/default.iprof</buildArg>
								<buildArg>-march=native</buildArg>
							</buildArgs>
						</configuration>
					</plugin>
				</plugins>
			</build>
		</profile>
		<profile>
			<id>monitored</id>
			<build>
				<plugins>
					<plugin>
						<groupId>org.graalvm.buildtools</groupId>
						<artifactId>native-maven-plugin</artifactId>
						<configuration>
							<imageName>demo-monitored</imageName>
							<buildArgs>
								<buildArg>--pgo=${project.basedir}/default.iprof</buildArg>
								<buildArg>-march=native</buildArg>
								<buildArg>--enable-monitoring=heapdump,jfr,jvmstat</buildArg>
							</buildArgs>
						</configuration>
					</plugin>
				</plugins>
			</build>
		</profile>
```

The same applies if I need to add SBOM, `-Os`, or any other customization while keeping a clean default version of the build for comparison.

**Comment by fniephaus** [comment]
> the above is already possible even without adding support for -DbuildArg=<...> by using the NATIVE_IMAGE_OPTIONS environment variable.

That's correct. Nonetheless, the discoverability of this env var is relatively poor. In the context of the NBTs, we either need to document this more, or provide the ability to inject args via an user property. I would be in favor of a dedicated user property, for example `-Dnative-image.build.args="--pgo-instrument"`, similar to exec:exec's `-Dexec.args="-X myproject:dist"`, because that will feel most intuitive for Maven users.

**Comment by olpaw** [comment]
> I would be in favor of a dedicated user property, for example -Dnative-image.build.args="--pgo-instrument", similar to exec:exec's -Dexec.args="-X myproject:dist", because that will feel most intuitive for Maven users.

Adding support for something like `-Dnative-image.build.args=<extra-native-image-args>` to the `native-maven-plugin`  is fine for me. I was not suggesting it to be a bad idea. I was trying to make an argument that ways to influence the behavior of a tool for the sake on convenience/debugging are common and useful.

**Comment by fniephaus** [comment]
> I was not suggesting it to be a bad idea.

And I was not suggesting that. ;) I think providing users with different ways to achieve the same is not always bad, at least they can choose what they want to use.

**Comment by galderz** [comment]
Not sure I understand the problem here. It is possible to pass in `buildArgs` via command, I've been doing that by just passing in `mvn ... -DbuildArgs=option1,option2,option3` without any issues, e.g.

```
mvn ... -DbuildArgs=-H:-DeleteLocalSymbols,-H:+SourceLevelDebug
```

What I find difficult is to append to existing `buildArgs` in pom.xml definitions.

Take `--bundle-create` as an example. There is no high level configuration option in the Maven plugin to pass that in (missing feature in native-build-tools integration?). So you have to fallback on `buildArgs`. Now, let's say that you always want to produce the bundle, so you decide to add it to the pom.xml:

```
<buildArgs>
    <buildArg>--bundle-create</buildArg>
</buildArgs>
```

Now, let's say you're doing some debugging and you want to avoid deleting local symbols, something you only want to do sporadically. The obvious thing to do is to pass in `-DbuildArgs=-H:-DeleteLocalSymbols` in the command line, but when you do that, the native image build no longer includes `--bundle-create` because all the additional build args passed in have been replaced by what gets passed in via the command line.

One might think you can workaround this by using an argument file, but that doesn't work here because `--bundle-create` can only be passed in via the command line and you get an error when trying to add it to `Args` in a `native-image.properties` file.

**Comment by galderz** [comment]
There are ways to manufacture a new system property such that it then gets appended to `buildArgs` but it involves defining a different system property which is confusing.

**Comment by melix** [comment]
#683 was for Gradle, I'm not sure about Maven, however "standard" maven property overrides should work

**Comment by alina-yur** [comment]
@dnestoro I'm happy to test on Maven. Should those also work in Maven? https://github.com/graalvm/native-build-tools/blob/50adf64f39a21c1c8cb5ff67250ea1c6d5aa74ea/docs/src/docs/asciidoc/gradle-plugin.adoc#command-line-options
